### PR TITLE
Tx iterator small optimization 4083 v8

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/7089

Describe changes:
- some optimizations for tx iterations

#11508 with persistent storage in `FlowWorkerThreadData`

### CODE REVIEW

File: src/app-layer-parser.c [hunk 1]
```diff
@@ -919,6 +918,23 @@ void AppLayerParserTransactionsCleanup(Flow *f, const uint8_t pkt_dir)
     SCLogDebug("start min %"PRIu64, min);
     bool skipped = false;
     // const bool support_files = AppLayerParserSupportsFiles(f->proto, f->alproto);
+    uint32_t tofree_nb = 0;
+    if (total_txs > UINT32_MAX) {
+        SCFree(txl->tx_id_free_list);
```
Comment (catenacyber) [inline src/app-layer-parser.c:923]: Why does `AppLayerParserGetTxCnt` return a u64 ?

**Comment by catenacyber** [comment]
Conflicts with https://github.com/OISF/suricata/pull/11518 which removes app-layer-rfb.c

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.

ERROR: QA failed on SURI_TLPR1_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 138 | 144 | 104.35% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 648 | 695 | 107.25% | 

Pipeline 21663

File: src/app-layer-parser.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-parser.c:923]: It's not named correctly, it should be something like `GetMaxId`

File: src/app-layer-parser.c [hunk 2]
```diff
@@ -919,6 +918,23 @@ void AppLayerParserTransactionsCleanup(Flow *f, const uint8_t pkt_dir)
...
+        txl->tx_id_free_list = NULL;
```
Comment (victorjulien) [inline src/app-layer-parser.c:924]: it doesn't seem to be necessary to free it here?

File: src/app-layer-parser.c [hunk 3]
```diff
@@ -919,6 +918,23 @@ void AppLayerParserTransactionsCleanup(Flow *f, const uint8_t pkt_dir)
...
+        txl->max_tx = 0;
+    } else if (txl->max_tx < total_txs) {
+        uint64_t *tmp = SCRealloc(txl->tx_id_free_list, total_txs * sizeof(uint64_t));
```
Comment (victorjulien) [inline src/app-layer-parser.c:927]: can we just double the size every time we need more space? No need to realloc for every additional tx in a flow then. Could just increase to the next power of 2 that fits the number of txs we need.

**Review by victorjulien — CHANGES_REQUESTED** [review]
see inline

File: src/app-layer-parser.c (hunk 1 above)
Comment (catenacyber) [reply src/app-layer-parser.c:923]: Oh...

So tx count is bounded by MaxId - MinId...

File: src/app-layer-parser.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-parser.c:924]: Ok for not freeing.

But we still need to handle this case where we have way too many live transactions (4 billion transactions).
Adding a new local bool to do that.

File: src/app-layer-parser.c (hunk 3 above)
Comment (catenacyber) [reply src/app-layer-parser.c:927]: ok, increasing by a step of 256

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11555

**Pull request closed without merging**